imxrt-hal/src/adc.rs

```diff
@@ -223,6 +223,205 @@ where
+where
+    ADCx: adc::ADC + AdcDmaSource,
+    P: Pin<ADCx>,
+{
+    /// Create a DMA Source object tieing the ADC to the given pin.
+    pub fn new(adc: ADC<ADCx>, pin: AnalogInput<ADCx, P>) -> Self {
+        AdcSource { adc, pin }
+    }
+}
+
+impl<ADCx, P> crate::dma::peripheral::Source<u16> for AdcSource<ADCx, P>
+where
+    ADCx: adc::ADC + AdcDmaSource,
+    P: Pin<ADCx>,
+{
+    type Error = ();
+
+    const SOURCE_REQUEST_SIGNAL: u32 = ADCx::SOURCE_REQUEST_SIGNAL;
+
+    fn source(&self) -> *const u16 {
+        &self.adc.reg.R0 as *const _ as *const u16
+    }
+
+    fn enable_source(&mut self) -> Result<(), Self::Error> {
+        let channel = <P as Pin<ADCx>>::Input::U32;
+        ral::modify_reg!(ral::adc, self.adc.reg, GC, ADCO: 1, DMAEN: 1);
+        ral::modify_reg!(ral::adc, self.adc.reg, HC0, |_| channel);
+        Ok(())
+    }
+
+    fn disable_source(&mut self) {
+        ral::modify_reg!(ral::adc, self.adc.reg, GC, ADCO: 0, DMAEN: 0);
+    }
+}
+
+/// StreamingAdc wraps AdcSource and the DMA machinery to provide a nice
+/// user interface for reading regular samples from the ADC.
+///
+/// This implementation uses two linear buffers and switches between them,
+/// yielding 256 samples at a time.
+///
+/// # Example
+///
+/// ```no_run
+/// static ADC_BUFFER: hal::dma::Buffer<[u16; 256]> = hal::dma::Buffer::new([0; 256]);
+/// static ADC_BUFFER2: hal::dma::Buffer<[u16; 256]> = hal::dma::Buffer::new([0; 256]);
+/// static mut streaming_adc: Option<hal::adc::StreamingAdc<hal::iomuxc::adc::ADC1, t41::P24>> = None;
+///
+/// #[interrupt]
+/// fn DMA8_DMA24() {
+///     unsafe {
+///         if let Some(adc) = &mut streaming_adc {
+///             adc.handle_interrupt();
+///         }
+///     }
+/// }
+///
+/// ...
+///
+/// let (adc1_builder, _) = adc.clock(&mut ccm.handle);
+/// let mut adc1 = adc1_builder.build(
+///     adc::ClockSelect::default(),
+///     adc::ClockDivision::default()
+/// );
+/// let mut a1 = adc::AnalogInput::new(pins.p24);
+///
+/// let mut channel = dma_channels[8].take().unwrap();
+/// channel.set_interrupt_on_completion(true);
+/// unsafe {
+///     streaming_adc = Some(hal::adc::StreamingAdc::new(
+///         adc1,
+///         a1,
+///         channel,
+///         &ADC_BUFFER,
+///         &ADC_BUFFER2
+///     ));
+///     cortex_m::peripheral::NVIC::unmask(interrupt::DMA8_DMA24);
+/// }
+///
+/// ...
+///
+/// // Inside the main loop
+/// let mut sadc = unsafe { streaming_adc.as_mut().unwrap() };
+/// sadc.poll(|buf| {
+///     adc_min = *buf.iter().min().unwrap();
+///     adc_max = *buf.iter().max().unwrap();
+/// });
+/// ```
+pub struct StreamingAdc<ADCx, P> {
+    buffer0: &'static crate::dma::Buffer<[u16; 256]>,
+    buffer1: &'static crate::dma::Buffer<[u16; 256]>,
+    peripheral: crate::dma::Peripheral<
+        AdcSource<ADCx, P>,
+        u16,
+        crate::dma::Linear<u16>,
+        crate::dma::Linear<u16>,
+    >,
+    unhandled_buffer_count: core::sync::atomic::AtomicUsize,
```

Comment:
Since `unhandled_buffer_count` is always modified in `&mut` methods, could this be simplified to a normal `usize`?

Reply:
My concern is since it gets modified from the interrupt, it doesn't _really_ have an exclusive borrow to the object.

Not sure what the Rust-y way to fix that is. Maybe make both these methods `&self` methods, and use an `UnsafeCell`?

Reply:
Of course, using `Circular` might circumvent this, since then we shouldn't need an ISR to fiddle with the buffers at all, there's only the user poll()ing as fast as they need.

Reply:
For a `StreamingAdc` with mutable methods shared with an ISR, I think we could guarantee exclusive mutable access using a combination of [`Mutex`](https://docs.rs/cortex-m/0.7.2/cortex_m/interrupt/struct.Mutex.html), `RefCell`, and `Option`.

- `Mutex` for exclusion between polling loop and ISR
- `RefCell` for interior mutability required by `Mutex`
- `Option` because it needs setup in `main()` before it's stuffed into a `static`

The types might be gnarly! But there's no unsafe, and it makes sure we have the exclusive access we say we need with `&mut self` methods.

```rust
use core::cell::RefCell;
use cortex_m::interrupt::Mutex;

type MyAdc = Mutex<RefCell<Option<StreamingAdc>>>;
static MY_STREAMING_ADC: MyAdc = Mutex::new(RefCell::new(None));

#[cortex_m_rt::interrupt]
fn DMA8_DMA24() {
    cortex_m::interrupt::free(|cs| {
        // Mutex<RefCell<...>> -> RefCell<...>
        let adc = MY_STREAMING_ADC.borrow(cs);
        // RefCell<Optional<...>> -> &mut Optional<...>
        let mut adc = adc.borrow_mut();
        // &mut Optional<StreamingAdc> -> &mut StreamingAdc
        let mut adc = adc.as_mut().unwrap();
        // Swap buffers, etc...
    });
}

fn main() {
    // Setup ADC n'at...

    // Prepare shared state between ISR, polling loop...
    cortex_m::interrupt::free(|cs| {
        let adc = MY_STREAMING_ADC.borrow(cs);
        let mut adc = adc.borrow_mut();
        *adc = Some(the_adc);
    });

    // Polling loop...
    cortex_m::interrupt::free(|cs| {
        // As before...
    });
}
```